lib/Basic/ParseableOutput.cpp

```diff
@@ -0,0 +1,484 @@
+        Inputs.push_back(CommandInput(OutFiles[BJAction->getInputIndex()]));
+      } else {
+        for (llvm::StringRef FileName : OutFiles) {
+          Inputs.push_back(CommandInput(FileName));
+        }
+      }
+    }
+
+    // TODO: set up Outputs appropriately.
+    file_types::ID PrimaryOutputType = Cmd.getOutput().getPrimaryOutputType();
+    if (PrimaryOutputType != file_types::TY_Nothing) {
+      for (llvm::StringRef OutputFileName :
+           Cmd.getOutput().getPrimaryOutputFilenames()) {
+        Outputs.push_back(OutputPair(PrimaryOutputType, OutputFileName.str()));
+      }
+    }
+    file_types::forAllTypes([&](file_types::ID Ty) {
+      for (auto Output : Cmd.getOutput().getAdditionalOutputsForType(Ty)) {
+        Outputs.push_back(OutputPair(Ty, Output.str()));
+      }
+    });
+  }
+
+  void provideMapping(swift::json::Output &out) override {
+    Message::provideMapping(out);
+    out.mapRequired("command", CommandLine); // Deprecated, do not document
+    out.mapRequired("command_executable", Executable);
+    out.mapRequired("command_arguments", Arguments);
+    out.mapOptional("inputs", Inputs);
+    out.mapOptional("outputs", Outputs);
+  }
+};
+
+class DetailedInvocationBasedMessage : public InvocationBasedMessage {
+  ArrayRef<const char *> Args;
+  std::string Executable;
+  SmallVector<std::string, 16> Arguments;
+  std::string CommandLine;
+  SmallVector<CommandInput, 4> Inputs;
+  SmallVector<OutputPair, 8> Outputs;
+
+public:
+  DetailedInvocationBasedMessage(StringRef Kind,
+                                 const CompilerInvocation &Invocation,
+                                 const InputFile &PrimaryInput,
+                                 ArrayRef<const char *> Args)
+      : InvocationBasedMessage(Kind, Invocation), Args(Args) {
+    // Command line and arguments
+    Executable = Invocation.getFrontendOptions().MainExecutablePath;
+    CommandLine += Executable;
+    for (const auto &A : Args) {
+      Arguments.push_back(A);
+      CommandLine += std::string(" ") + A;
+    }
+
+    // Primary Input only
+    Inputs.push_back(CommandInput(PrimaryInput.getFileName()));
+
+    // Output for this Primary
+    auto OutputFile = PrimaryInput.outputFilename();
+    Outputs.push_back(OutputPair(file_types::lookupTypeForExtension(
+                                     llvm::sys::path::extension(OutputFile)),
+                                 OutputFile));
+
+    // Supplementary outputs
+    const auto &primarySpecificFiles = PrimaryInput.getPrimarySpecificPaths();
+    const auto &supplementaryOutputPaths =
+        primarySpecificFiles.SupplementaryOutputs;
+    supplementaryOutputPaths.forEachSetOutput([&](const std::string &output) {
+      Outputs.push_back(OutputPair(file_types::lookupTypeForExtension(
+                                       llvm::sys::path::extension(output)),
+                                   output));
```

Comment:
I'm not sure it's a good idea to use `lookupTypeForExtension()` like this. A build system could pass filenames with no extension or the wrong extension; if they do, we should still treat these correctly, using a file type based on which field in the supplementary output path field we found the path in. (This probably means `forEachSetOutput` would pass the file type to the lambda.)

Reply:
This is a great observation, thank you. We do indeed have all the information we need to determines this based on action and known file type instead of inferring it. I modified `forAllOutputPaths` to pass along the filetype.